### PR TITLE
Disable edit when scheduled to be published

### PR DESCRIPTION
**Pull request by flamerohr**
This is to prevent further changes made to the draft (which users could "Save Draft") that is embargoed, so that those changes do not bypass the approval process.

This pull request works better with https://github.com/silverstripe-australia/silverstripe-advancedworkflow/pull/267
So you'd be able to disable edit, then someone will permissions would be able to cancel the embargo to allow further editing.

### CODE REVIEW

File: code/extensions/WorkflowEmbargoExpiryExtension.php [hunk 1]
```diff
@@ -358,4 +359,26 @@ public function setIsWorkflowInEffect() {
 	public function getIsWorkflowInEffect() {
 		return $this->isWorkflowInEffect;
 	}
+
+    /**
+     * Add edit check for when publishing has been scheduled and if any workflow definitions want the item to be disabled.
+     */
+    public function canEdit($member) {
+        $disabled = false;
+        $definitions = $this->workflowService->getDefinitionsFor($this->owner);
```
Comment (nyeholt) [inline code/extensions/WorkflowEmbargoExpiryExtension.php]: Can you add a check for where `this->owner->PublishOnDate` is actually defined before querying the workflow definitions?

**Comment by nyeholt** [comment]
I like the idea, I'm just not sure whether the 'DisableBeforeEmbargo' flag should exist on the definition or the object itself, as I could see the functionality being useful outside the context of a full approval workflow being applied. @tractorcow any input ?

Separately - could the commits be squashed down ?

File: code/extensions/WorkflowEmbargoExpiryExtension.php (hunk 1 above)
Comment (flamerohr) [inline code/extensions/WorkflowEmbargoExpiryExtension.php]: Good idea, that'd make sense to check that first

**Comment by flamerohr** [comment]
I was contemplating about moving the flag to a permissions for groups and users.
What do you reckon?
Seems to simplify some of the checking as well.

**Comment by flamerohr** [comment]
I've moved the `WorkflowDefinition.DisableEditEmbargoed` to a Provided permission for users/groups/roles `EDIT_EMBARGOED_WORKFLOW`, the implementation is cleaner than previously.

(Have squashed things to one commit as well.)

**Comment by nyeholt** [comment]
I'm not sure that only having a global canEditInEmbargo is necessarily the best approach, compared to the object/page having the flag on it directly. It's probably reasonable to have it as a secondary 'override' regardless of whatever is set on an object directly, but I think it'd be good to have a flag as part of WorkflowEmbargoExpiryExtension as to whether editing should be allowed - this can then be set by the PublishItemWorkflowAction.

**Comment by flamerohr** [comment]
I'm not sure I understand what you mean, so do you mean something like:
* Have a flag on `WorkflowEmbargoExpiryExtension` to control editing
* During the `PublishItemWorkflowAction::execute()` it could set that flag to be editable or not?

**Comment by nyeholt** [comment]
Essentially yes. By just having the EDIT_EMBARGOED_WORKFLOW, it means that any user with that permission can perform that action on any embargoed content. Whereas I think it makes more sense for that to be a per-item decision as to whether it's editable or not while embargoed (though having the global perm does mean a 'power' user type role could override the per-item behaviour). 

To take a completely different look at it - maybe another way of doing things would be that rather than a workflow ending immediately (and the future publish date being set), the workflow instance itself remains active on the content while it's embargoed, so that the workflow settings would determine whether something was editable; when published, the workflow would be subsequently completed. This does pretty drastically change the way 'embargo' works though, particularly around a future unpublish date.

**Pull request merged**